nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java

```diff
@@ -135,71 +135,26 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
     boolean newTable = base == null;
     String newMetadataLocation = writeNewMetadataIfRequired(newTable, metadata);
 
-    String refName = client.refName();
-    boolean failure = false;
+    AtomicBoolean failure = new AtomicBoolean(false);
     try {
       String contentId = table == null ? null : table.getId();
       client.commitTable(base, metadata, newMetadataLocation, contentId, key);
-    } catch (NessieConflictException ex) {
-      failure = true;
-      if (ex instanceof NessieReferenceConflictException) {
-        // Throws a specialized exception, if possible
-        maybeThrowSpecializedException((NessieReferenceConflictException) ex);
+    } catch (NessieConflictException | NessieNotFoundException | HttpClientException ex) {
+      NessieUtil.handleExceptionsForCommits(ex, client.refName(), failure);
+    } catch (NessieBadRequestException ex) {
```

Comment:
Here too, I think this catch block is a code smell, but this one is trickier to solve.

The issue is that in the test called `createTableViaTransactionThatAlreadyExistsAsView`, we create a transaction based on current hash (let's call it c0), then we commit something outside the transaction (let's call it c1), then we attempt to commit the transaction.

As a result, the transaction/client is trying to commit using the expected hash of c1, but it should really use c0 as the expected commit hash.

There is a property to convey the commit ID: `NESSIE_COMMIT_ID_PROPERTY`, but it doesn't work when creating things, only when updating.

I *think* we could solve the problem by storing the current expected hash in a field inside `NessieTableOperations`, then passing that to the client:

```java
public class NessieTableOperations extends BaseMetastoreTableOperations {
...
  private Reference reference;
...
  NessieTableOperations(...) {
    ...
    reference = client.getRef().getReference();
  }
...
  protected void doRefresh() {
    ...
    String metadataLocation = null;
    this.reference = client.getRef().getReference();
    ...
  }
...
  protected void doCommit(TableMetadata base, TableMetadata metadata) {
      ...
      client.commitTable(base, metadata, newMetadataLocation, contentId, key, reference);
      ...
  }
}
```

I did a quick experiment with the above suggestion and the test threw `NessieConflictException` as expected, instead of `NessieBadRequestException`.

@dimas-b what would you suggest in this situation?

Reply:
In my understanding, commits will be retried on failure for `NessieConflictException` and it will call `doRefresh()` for retry which will again use the new reference?

Reply:
I think the proper way to fix it is to check if any different kind of content exist for the key while committing this key in `doCommit()` of `NessieTableOperation` and `NessieViewOperation`. But that is an extra round trip with server and might degrade commit performance slightly?

Reply:
1. Conflicts will _not_ be retried for `NessieReferenceConflictException` since we throw specialized errors, see here: https://github.com/apache/iceberg/blob/89b3a2111968f0aed79ec0ce1c87e2035e0d7f7c/nessie/src/main/java/org/apache/iceberg/nessie/NessieUtil.java#L188-L192
2. My main point was: we should get rid of this `catch (NessieBadRequestException ex)` block as, in my opinion, it denotes a code smell.

Reply:
> I think the proper way to fix it is to check if any different kind of content exist for the key while committing this key in doCommit() of NessieTableOperation and NessieViewOperation. But that is an extra round trip with server and might degrade commit performance slightly?

This solution is viable but would imho be less efficient since, as you mentioned, it involves an extra roundtrip. 

I personally think that if `NessieTableOperation` could "memorize" the commit hash of that last time it did a refresh, then reuse it to commit the table changes, that would solve the problem in a better way since it does not involve fetching contents.

Reply:
That particular testcase is tricky because `createTransaction()` will call `refresh()` which already has code to check if any other content exist with the same name. But during `commitTransaction()` it will not call `refresh()` so that code is never hit and we need to check again if any new content created with same key concurrently. 

So, I have added that logic now instead of catching `NessieBadRequestException` later on.

I am not sure about `memorize` commit hash solution. Last time (some other PR) we concluded that we need to be as stateless as possible. Lets see what others think.

Reply:
I think it's a good idea to record the current commit hash for "new" table/view commits at the time `NessieTableOperations` is created. At the same time, if the commit is an update, I think we should use the hash from the metadata (basically track it for views too).

Reply:
Re: `NessieBadRequestException`, I agree that we should not try to "handle" it. Instead the catalog should be implemented such that `NessieBadRequestException` does not happen.